### ISSUE TITLE
Why use mean_absolute_error as loss function instead of mean_squared_error?

### ISSUE THREAD

**Issue by IUsedToBeAPygmy** [deepfakes/faceswap#120]
I've always been taught MSE would be the standard loss function to pick.
On the other hand, I don't know what most of the other Keras loss functions do exactly either.

Can somebody explain the reasoning for choosing MAE?

And while I'm at it, can somebody explain (or point me to some reading material)  why the Adam optimizer is chosen over any of the others?

**Comment by shaoanlu** [comment]
I have very limited knowdge about autoencoders, but I’ve seen both L1 and L2 loss benn used in papers (as the loss func. to ground truth image). I think its somehow hurustic and empirical. In fact, I’ve tried both losses in this project but didn’t find much difference in terms of output quality.

**Comment by gdunstone** [comment]
@IUsedToBeAPygmy im experimenting with using the Nadam optimizer due to this paper:

https://openreview.net/pdf?id=OM0jvwB8jIp57ZJjtNEZ

**Comment by gdunstone** [comment]
![image](https://user-images.githubusercontent.com/4244600/35817518-ba212d0c-0af1-11e8-8c69-7f151e6181fd.png)

Quick test.

**Comment by IUsedToBeAPygmy** [comment]
@gdunstone - I've just tried to replace the used Adam with Keras' standard Nadam, but all I get is completely red images so apparently it's not just a drop-in replacement.
Did you get it to work and if so, how?

**Comment by IUsedToBeAPygmy** [comment]
Also, I've created a model that implements the loss function as a mix of SSIM and L2  and uses PELU instead of LeakyReLu which seems to work rather well. If people want I can put in a pull request.

http://research.nvidia.com/sites/default/files/pubs/2017-03_Loss-Functions-for/NN_ImgProc.pdf

**Comment by gdunstone** [comment]
@IUsedToBeAPygmy Thanks! Can you work it as a plugin similar to how GAN is being used?

I literally did a drop in replacement of Adam with Nadam, and the preview output while I was gathering the above data looked totally fine.

**Comment by IUsedToBeAPygmy** [comment]
I'm using the SSIM from keras.contrib and mix its result with the result of a normal loss function like mean_absolute_error (using a very small factor for the SSIM).

https://github.com/keras-team/keras-contrib/blob/master/keras_contrib/losses/dssim.py

Keras.contrib also supplies alternative advanced activation functions like PELU.

**Comment by gdunstone** [comment]
Can you get us some of those sweet sweet numbers? Or hook us up with a hit of graphs? :stuck_out_tongue:

**Comment by IUsedToBeAPygmy** [comment]
I'm going to hook you up with the code soon so you can make your own graphs :)
